### PR TITLE
[REF] Modularize io

### PR DESCRIPTION
**Pull request by jbteves**
Closes None .

Changes proposed in this pull request:

- Switches `outdir` from a function argument to an `io` global
- Adds `--prefix` option to prepend a name to the output, also an `io` global
- Adds `--convention` option to specify `bids` or `kundu` output, with `bids` the default
- Modifies smoke test for `test_io` to match these changes

Note: the current solution I have is a little wonky for naming the output files, but I think it should be okay. `split` at the end denotes something that has formatting according to echo-number, which is something that `split_ts` requires in order to function properly in the way I wrote this. This is a draft PR so I'm happy to make loads of modifications, but I think the spirit of this implementation should work to solve the problem of varying naming conventions over time while retainining backwards compatibility.

This current draft does *not* allow backwards compatibility to keep `.json` files instead of `.tsv` files.

### CODE REVIEW

**Comment by jbteves** [comment]
Hm, an interesting problem I'm running into is that when I run
`pytest test_t2smap.py` it actually passes, but if I run `pytest .` in the test directory, that same file fails. I have no idea why that's happening, but it seems to be related to directory creation of some kind.

**Comment by jbteves** [comment]
I'm still going to take a look at what's causing this strange behavior for the tests, but @tsalo if you take a look at the table and the function that uses it to generate filenames, I think you'll get the gist of what this PR does and you can decide if you like the implementation or not. Maybe @dowdlelt could also take a quick look?

File: tedana/io.py
Comment (dowdlelt) [inline tedana/io.py:21]: These aren't going to override user input are they? I've looked around and this pull request is definitely what I was imagining. And bonus, if some new naming convention takes the world by storm - just add another set at the end and index with [2]!

Comment (jbteves) [reply tedana/io.py:21]: No, this is a just a default. In the main workflow it's specified with `--convention`, and in the API it's specified with `convention=`

Comment (jbteves) [reply tedana/io.py:21]: > if some new naming convention takes the world by storm - just add another set at the end and index with [2]
That's what I was thinking. Plus at some point there will be BIDSv2.0 anyway. So these convention changes are probably inevitable.
If we want to get fancy we can set things up so that any given file has a full description and some other features, but I think this is an okay start.

**Comment by dowdlelt** [comment]
In general I think this looks great, and it is what I was imagining - a system ask minimal changes from the user (specific convention, if you want), but is flexible enough that code remains legible without a litany of if statements. solid :+1: from me.

File: tedana/io.py
Comment (handwerkerd) [reply tedana/io.py:21]: The same line confused me. This might be a good place for a comment to note `convention = 'bids'` sets the default, which is typically changed when tedana.py is run.

**Comment by handwerkerd** [comment]
As @dowdlelt notes, the current output names are  "kundu" and "bids" I think this is doubly problematic because "kundu" isn't particularly informative and "bids" assumes the bids derivatives outputs won't ever change (i.e. we won't end up with bids_v1, bids_v2, etc).

My proposal is to have 3ish options: "orig"(I'm ok with kundu), "bids" and "version X". "bids" will always output the most recent version of the bids derivatives output. "version X" will output whatever was the current output for the given version number. That is any version number less than or equal to 0.0.9 would be identical to "orig", but if bids changes in the future, using the same version number as was used for an older execution will produce matching outputted file names. 

I suggest using only released versions. That is, if v0.0.9+44 is different from v0.0.9+45, that difference will be lost with time. We'll only commit to saving file name changes that made it into a unique release.

To do this, the `if` clause at https://github.com/jbteves/tedana/blob/a5490a053d4f0cd5d7031e8a9d95f4488d53c931/tedana/io.py#L124-L128 would need to become _slightly_ more complex.
In io.py, `img_table` would need to have a new element for each table index, which might get messy, if there are too many version changes. I think we can punt on that for now. If there are >3 file name versions being stored, we can change img_table to a structure where elements could be accessed by descriptive key and column label rather than a descriptive key and an index number.

Semi-related, particularly if a non-default `--convention` is being used, that information should be mentioned in the tedana output.

Thoughts?

**Comment by handwerkerd** [comment]
The json & tsv files always use bids file naming conventions. I previously suggested that we stick BIDS conventions for file type (i.e. json vs tsv), but we do allow for the file name to chance with the options. Do others have opinions on this?

**Comment by jbteves** [comment]
That all makes sense. I would propose the following, then:
1. Make a clarifying comment that `convention` will be overridden by the workflow functions.
2. Make the existing `img_table` a nested dict, wherein we have the image type as the outer key and the naming convention as an inner key
3. Make the `bids` inner key point to the latest bids version, while adding a name indicating the current `bids` derivatives version (will need @tsalo's input on that).
4. I realized that I never added table naming; I'll do something similar to the work on the `img_table` where we generate a table name.

Apparently I forgot to hit "comment" on this when I wrote it earlier this afternoon.

I think we should have a table of table-names so that we can switch conventions for the names of tables as well. It would be weird to only do images.

**Comment by jbteves** [comment]
Tests are passing locally and so did the linter, so this is ready for review. I'm pretty open to making minor changes on the function names. In a future PR, I will add functionality to add the entire file manifest to `dataset_description.json` so that you can easily query the dataset even with prefixes.

**Comment by handwerkerd** [comment]
It's crashing when I'm trying to run with `--convention kundu` Which is what is currently listed in the help instructions
```
2 voxels in user-defined mask do not have good signal. Removing voxels from mask.
Traceback (most recent call last):
  File "/opt/anaconda3/bin/tedana", line 33, in <module>
    sys.exit(load_entry_point('tedana', 'console_scripts', 'tedana')())
  File "/Users/handwerkerd/code/tedana_community/master/tedana/tedana/workflows/tedana.py", line 816, in _main
    tedana_workflow(**kwargs)
  File "/Users/handwerkerd/code/tedana_community/master/tedana/tedana/workflows/tedana.py", line 504, in tedana_workflow
    io.filewrite(masksum, 'adaptive mask', ref_img)
  File "/Users/handwerkerd/code/tedana_community/master/tedana/tedana/io.py", line 463, in filewrite
    root = gen_img_name(img_type, echo=echo)
  File "/Users/handwerkerd/code/tedana_community/master/tedana/tedana/io.py", line 68, in gen_img_name
    format_string = img_table[img_type][convention]
KeyError: 'kundu'
```
It looks to me like there are a few issues. The main one is that, in constants.py, the options are now `allowed_conventions = ('orig', 'bidsv1.5.0')` but the input options are `kundu` and `bids` I think the plan was to allow some conversation for convention names (i.e. bids & kundu conversion to specific version numbers, but I don't see where that's happening.
Also, io.py has a nice `check_convention` function, but I'm not sure if it's every actually called, which is why the crash isn't more descriptive.

**Comment by jbteves** [comment]
Oops, I forgot to switch the help text after you recommended switching to `orig`. I'll fix that momentarily.
`allowed_conventions` is still accurate, it's just that now `bids` is a variable that points to `bidsv1.5.0`, rather than making a duplicate entry in the dicts.

**Comment by jbteves** [comment]
@handwerkerd did you see anything else that I should address?

**Comment by handwerkerd** [comment]
Several functions in io.y: write_split_ts, & writefeats, writeresults, & writeresults_echoes have parameters: out_dir & prefix removed. Those parameters are now stored globally. There is an additional global parameter, convention, and several constants are stored in constants.py. I think this is all a good thing, but if someone stumbles onto this code, it might be a bit confusing to figure out where variables are defined.

Would it make sense to explain the globals in help text for each function so, if someone tries to run them as modular functions they can figure out what is happening without too much digging?

This could also be explained in the module header text ( https://github.com/jbteves/tedana/blob/1135968871d276426621548b67d8b8f6e31c6b38/tedana/io.py#L2 ) but I'm not sure if the convention is to keep that very short.

**Comment by jbteves** [comment]
@handwerkerd does 01ceae9 address your documentation concerns?

File: tedana/io.py
Comment (handwerkerd) [inline tedana/io.py:610]:
```suggestion
        )
        

def set_convention(in_convention):
    """
    Parses user inputted convention to a subset of file naming conventions
    that are used in constants.py

    Parameters
    ----------
    in_convention : :obj:`str`
        A file naming convention inputted by the user
        Currently this is 'orig', 'bids' or a tedana version number, 
        like 0.0.9. If a version number is entered


    Returns
    -------
    convention : :obj:`str`
        A convention that is used in constants.py is globally set
        within this function
    """

    from packaging import version
    from .constants import bids, allowed_conventions

    if version.parse(in_convention).release:
        # If the convention is a version number
        # identify the correct convention for the tedana version
        version_num = version.parse(in_convention)
        if version_num <= version.parse("0.0.9"):
            convention = 'orig'
        else:
            convention = bids
    elif in_convention.casefold() == 'bids':
        # If the convention is 'bids' use the newest version
        # of the bids naming convention
        convention = bids
    else:
        # Otherwise, just assign the inputted convention and raise
        # error if the convention is not valid
        convention = in_convention
        check_convention()

    return convention

```

File: tedana/workflows/tedana.py
Comment (handwerkerd) [inline tedana/workflows/tedana.py:353]:
```suggestion
    io.convention = io.set_convention(convention)
```

I created a new function, io.set_convention so that more logic can be used to parse user inputted convention options into a smaller list of options that are used in constants.py. Right now, it includes the current two options, plus a way to input a tedana version number, which would then convert it to the correct convention name. Note that the choices in argparse won't allow for a version number to be entered & I'm not sure how to allow both fixed choices and a more flexible choice within argparse. This is current irrelevant, since there are only two options for conventions, but I figured it's better to modularization this now to allow for easier expansion later on.

**Review by handwerkerd — CHANGES_REQUESTED** [review]

**Comment by jbteves** [comment]
@handwerkerd and @dowdlelt do you mind taking a look at this latest version? I know that @tsalo would really, really like this merged in soon so that he doesn't have to keep maintaining the BIDS branch.
Of note to @emdupre and @eurunuela as well is that this branch would add a new module `constants.py` where we can define constants, and that the `io.py` module has a complete module docstring in this branch.

File: tedana/decomposition/pca.py
Comment (handwerkerd) [inline tedana/decomposition/pca.py:152]: This PR removes the file names from this description. I'd lean towards keeping those file names in  and maybe change the header to "Default Filename"
The benefit of removing the file name is, if we keep them, that means, every time we change file names, we'll also need to change the defaults in each of these help texts, but I think the clarity benefits of including the file names is still more useful.

File: tedana/workflows/tedana.py
Comment (handwerkerd) [inline tedana/workflows/tedana.py:349]: You have a problem that bothered me earlier. You are running this functions in `io` and assigning the new value to `convention` and `prefix` in those functions Those new values do not seem to leave those functions. When I suggested a change here, I had: `io.convention = io.set_convention(convention)` and that function ended with `return convention` I don't know if there is a better way to do it, but that worked and what you have doesn't. If you run the workflow with `--convention orig` it still runs with the default bids.

**Review by handwerkerd — CHANGES_REQUESTED** [review]

File: tedana/workflows/tedana.py
Comment (notZaki) [reply tedana/workflows/tedana.py:349]: Try [inserting the global keyword](https://github.com/notZaki/tedana/commit/4ec51d717c4bb34dc99729e935cc7b8755d7f9e3) in `set_function`.

**Review by tsalo — CHANGES_REQUESTED** [review]
I have some initial thoughts. Some of them are higher level, while others are just typos in the docstrings.

File: tedana/io.py
Comment (jbteves) [reply tedana/io.py:2]: Whoops, didn't realize that would render that way. Do you think it would be good to just ditch what's basically a table of contents up top, and keep it to comments above each section?

File: tedana/constants.py
Comment (jbteves) [reply tedana/constants.py:1]: I'd thought about doing that but figured the implementation would take longer; would you rather open an issue to do that later and keep this, or have me go ahead and add that before merging?

Comment (tsalo) [reply tedana/constants.py:1]: We can handle it in #691 if you'd rather not do it here.

File: tedana/workflows/tedana.py
Comment (jbteves) [reply tedana/workflows/tedana.py:85]: Hm, I wanted it for my own purposes as well. Do you think it's harmful to add it?
You can't make a prefix with `out_dir` because that'll create a directory with the prefix name in it.

Comment (tsalo) [reply tedana/workflows/tedana.py:85]: Maybe it would be useful outside of BIDS compatibility. Could you expand on your use-case?

Ultimately, I think this is a new argument and behavior, so we should probably float it by the group in its own issue. We talked about prefixes a _long_ time ago in #106, and decided against them, but the group was much smaller then and others may feel differently.

> Do you think it's harmful to add it?

I think there are ways to work around prefixed files in nipype/fMRIPrep, so, while it _could_ add a bit more work for BIDS Apps, I don't think it would be harmful _per se_.

Comment (jbteves) [reply tedana/workflows/tedana.py:85]: Hm, I can get rid of that for now. The short version is that I'm using AFNI, not BIDS Apps, but have files where I still have BIDS prefixes. So it would simplify some things for me as someone who's only partially in that ecosystem. Perhaps we could reopen #106 ?

Comment (handwerkerd) [reply tedana/workflows/tedana.py]: Reading #106, it looks to me like we tabled the prefix discussion to have for whenever we make the file naming more BIDS friendly (i.e. now). I was inclined to welcome something like prefix then & I still am. If everyone else things this is a relatively benign change, I'd prefer to keep it.

**Comment by handwerkerd** [comment]
I'm noticing that, when using the BIDS convention, `S0map.nii.gz` and `T2starmap.nii.gz` are still being created, in addition to `desc-full_S0map.nii.gz` and `desc-full_T2starmap.nii.gz`. For the T2starmap, I think the problem arises here: https://github.com/ME-ICA/tedana/blob/b2a4b093c5b79cdf826f420473fb875c4fd2e673/tedana/workflows/tedana.py#L466
I suspect this is a problem originating in #691 rather than this PR, but it could be addressed in either.

File: tedana/io.py
Comment (handwerkerd) [reply tedana/io.py:496]: @jbteves, This is minor, but I wanted to make sure you saw it. The parameter descriptions don't match the new filewrite function inputs.

File: tedana/workflows/tedana.py
Comment (tsalo) [reply tedana/workflows/tedana.py:85]: If we move forward with prefix, then I'd rather merge prefix with out_dir than add a new argument. Prefix could include a path, and as long as it ends with a separator, it can fully replace out_dir.

File: tedana/io.py
Comment (tsalo) [reply tedana/io.py:2]: I'd perhaps use `See Also` sections in the docstrings and/or directly incorporate this info into `api.rst`. I am not sure how Pythonic module-level comments are.

File: tedana/workflows/tedana.py
Comment (jbteves) [reply tedana/workflows/tedana.py:85]: That's going to complicate the logic a bit and I worry that it's slightly confusing to users (this both "names" and "places" the files, rather than "this one names" and "this one places"). Is there some reason we should avoid having two arguments? We already have a ton of parameters anyway, in my view.

File: tedana/constants.py
Comment (jbteves) [reply tedana/constants.py:1]: It's in-progress but apparently I've upset Python 3.5 here.

File: tedana/workflows/tedana.py
Comment (tsalo) [reply tedana/workflows/tedana.py:85]: `prefix` is AFNI-like, and I was under the impression that that was how the argument behaved in AFNI.

Comment (jbteves) [reply tedana/workflows/tedana.py:85]: `-prefix` is a common but very catch-all parameter that can be inconsistent in AFNI, used to variously do any of the following:
1. Place the output in a certain directory
2. Prefix a special name
3. Specify a complete name
4. Specify file type

However, I don't think we want to emulate having one option do all of those things here, particularly since we're enforcing certain naming conventions strictly. This may just be a hangup on my end so I'll ping @handwerkerd to give his thoughts, too.

**Comment by jbteves** [comment]
> I suspect this is a problem originating in #691 rather than this PR, but it could be addressed in either.
That is incorrect; I doubled the line to make sure I would rewrite it correctly, and forgot to delete a duplicate. Bafflingly, it should be triggering errors in the integration tests, though. I may have to dig to find out why that's not happening.

File: tedana/workflows/tedana.py
Comment (tsalo) [reply tedana/workflows/tedana.py:85]: Ah, I didn't realize it was so variable.

**Comment by notZaki** [comment]
Since Python 3.5 is being difficult in the tests, at what point will it be dropped from the tests? It's reached end of life, and some of the functionality (e.g. interactive reports) don't work on it.

**Comment by tsalo** [comment]
>Since Python 3.5 is being difficult in the tests, at what point will it be dropped from the tests? It has reached end of life, and some of the functionality (e.g. interactive reports) don't work on it.

@notZaki I like this idea, although I don't think we should bundle that into this PR. At minimum, we should start considering a deprecation timeline for it. Would you be willing to open an issue?

**Comment by jbteves** [comment]
@tsalo and @notZaki I have opened an issue already. Since we want to get this merged, I'll go ahead and work around it in this PR.

File: tedana/workflows/tedana.py
Comment (handwerkerd) [reply tedana/workflows/tedana.py:85]: I weakly lean toward keeping out_dir and prefix separate. I'm fairly sure, in AFNI, when prefix is used for a directory you need to explicitly notate that. That is: `--out_dir ./newdir/identifier` is easily understandable, but `--out_dir identifier` could mean add the file prefix `identifier` or put files in an output directory called `identifier` We could require `--out_dir ./identifier/` but that seems like a confusing change in functionality.

File: tedana/config/img_table.json
Comment (tsalo) [inline tedana/config/img_table.json:8]: T2starmap and S0map are full-on suffixes, rather than values associated with an entity (e.g., `desc`).

Comment (jbteves) [reply tedana/config/img_table.json:8]: So I should actually revert this change?

**Review by handwerkerd — APPROVED** [review]

File: tedana/io.py
Comment (tsalo) [inline tedana/io.py:505]: Adding a space here breaks the rendering. Compare `img_type` and `ref_img` in the RTD CI build:

image

**Comment by jbteves** [comment]
@tsalo the one thing that seems unresolved is API documentation, because the entire module-level docstring gets rendered on RTD. I feel like we need to retain the documentation so that things are clear to future developers (read: us 6 months from now), but obviously the rendering is unwieldy. Any thoughts on the best approach there?

Additionally, it would make the rendering much larger, but I could put the elements that are rendering on one line (mostly function names in various categories) as separate lines instead.

**Comment by handwerkerd** [comment]
@tsalo, This seems fairly close to done. Do you want any changes here before merging with #691 or do you feel comfortable merging with 691 & then having a general review before merging with Main?

**Comment by tsalo** [comment]
I will try to review later today and/or over the weekend. I mostly want to take a deeper look at the global variables before I approve.

**Review by tsalo — APPROVED** [review]
Approving now. We can evaluate some stuff (e.g., globals, jsons, and large docstrings) further within the main PR.

**Pull request merged**
